RFC-0004-expression-eval.md

```diff
@@ -0,0 +1,80 @@
+
+In this RFC, we propose to optimize row expressions as a whole, not just function calls.  This has a few advantages:
+
+1) It allows for constant folding to be applied to expressions which are not function calls, such as special form expressions.
+2) It can reduce the overhead of constant folding by allowing the evaluation engine to evaluate the expression as a whole, rather than evaluating each function call individually.
+
+### [Optional] Goals
+
+* Allow for a pluggable implementation of expression evaluation and use this in the Presto optimizer for constant folding
+* Allow connector authors to use this plugin to optimize their own expressions (for example, for filter pushdown)
+
+### [Optional] Non-goals
+
+## Proposed Implementation
+
+A new SPI will be added to Presto to allow for a pluggable implementation of expression evaluation.  A default implementation of this will be added which delegates to the existing expresion evaluation code, but this may be overridden by a custom plugin.
+
+A new plugin will be implemented which will optimize row expressions by serializing them and sending them to the Presto sidecar for evaluation.  The Presto sidecar will have a new endpoint which can take in a serialized row expression and return the result of evaluating it.
+
+> Endpoint: /v1/evaluate
+> 
+> HTTP verb: POST
+> 
+>Body: JSON list of serialized row expressions
+>
+> Response: JSON list of optimized serialized row expressions
+ 
+The new plugin will be integrated into the Presto optimizer through a new optimizer rule similar to `RowExpressionOptimizer`, except which uses the new SPI to evaluate the expressions.  There will be a feature toggle which can enable or disable the use of this new SPI, however eventually it will become mandatory to use it.
+
+## [Optional] Metrics
+
+This is a 0 to 1 feature, and its success will be largely dictated by two metrics:
+
+1) The performance of evaluating expressions, particularly large, complicated expressions.  This can be measured by comparing the time it takes to evaluate a query with and without the new SPI.
+2) The ability to perform constant folding in C++ clusters with custom functions which are only registered in C++.
+
+## [Optional] Other Approaches Considered
+
+A per-function approach was initially considered, however this is chosen in favor of that because not only is the API cleaner, we also expect that it can scale better.
+
+## Adoption Plan
+
+- This feature will be disabled by default.
+- After a period of time, the default will be to enable it.
+
+## Test Plan
+
+In addition to using the existing test coverage for expression evaluation to catch bugs in the expression optimizer, we will also use a fuzzer to generate random expressions and evaluate them with and without the new SPI to ensure that the results are the same.
```

Comment:
It would be nice to add a section about performance testing and also add e2e tests for the optimizer with this new functionality enabled. Specifically, it would be nice to make sure errors are handled properly, e.g. sidecar crashing, not responding, taking very long time to respond, not able to evaluate a function, etc.